CGEN bug with sumtypes when performing nested unwrappings

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 ed9d672, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 45fd7eb.ed9d672
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 5900X 12-Core Processor  
|Memory              |101.36GB/127.92GB
|                    |
|V executable        |S:\repo\vlang\v.exe
|V last modified time|2025-01-15 14:09:01
|                    |
|V home dir          |OK, value: S:\repo\vlang
|VMODULES            |OK, value: C:\Users\john3\.vmodules
|VTMP                |OK, value: C:\Users\john3\AppData\Local\Temp\v_0
|Current working dir |OK, value: S:\vSelenium
|                    |
|Git version         |git version 2.45.1.windows.1
|V git status        |weekly.2025.1-83-ged9d672e
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders, r2) 14.2.0
|clang version       |(built by Brecht Sanders, r2) clang version 19.1.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  testit.v && S:\vSelenium\testit.exe`
```v
import os {File, open_append}

type Foo = string | int | f32 | File

fn bar(log ?Foo) {
	if log != none {
		if log is string {
			logger := open_append(log) or {panic(err)}
		}
	}
}

fn main() {
	bar(none )
}
```

**What did you see?**
```
testit.v:8:4: warning: unused variable: `logger`
    6 |     if log != none {
    7 |         if log is string {
    8 |             logger := open_append(log) or {panic(err)}
      |             ~~~~~~
    9 |         }
   10 |     }
================== C compilation error (from tcc): ==============
cc: C:/Users/john3/AppData/Local/Temp/v_0/testit.01JHN8G6G432GK00MTZ5377SKH.tmp.c:4464: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/john3/AppData/Local/Temp/v_0/testit.01JHN8G6G432GK00MTZ5377SKH.tmp.c:6780: error: field not found: _main__Foo
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

error



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5025,11 +5025,17 @@ fn (mut g Gen) ident(node ast.Ident) {
 			if is_auto_heap {
 				g.write('(*(')
 			}
+			is_option = is_option || node.obj.orig_type.has_flag(.option)
 			if node.obj.smartcasts.len > 0 {
 				obj_sym := g.table.sym(g.unwrap_generic(node.obj.typ))
 				if !prevent_sum_type_unwrapping_once {
+					nested_unwrap := node.obj.smartcasts.len > 1
+					if is_option && nested_unwrap && obj_sym.kind == .sum_type {
+						g.write('*(')
+					}
 					for i, typ in node.obj.smartcasts {
-						is_option_unwrap := is_option && typ == node.obj.typ.clear_flag(.option)
+						is_option_unwrap := i == 0 && is_option
+							&& typ == node.obj.orig_type.clear_flag(.option)
 						g.write('(')
 						if i == 0 && node.obj.is_unwrapped && node.obj.ct_type_var == .smartcast {
 							ctyp := g.unwrap_generic(g.type_resolver.get_type(node))
@@ -5037,11 +5043,13 @@ fn (mut g Gen) ident(node ast.Ident) {
 						}
 						if obj_sym.kind == .sum_type && !is_auto_heap {
 							if is_option {
-								if !is_option_unwrap {
-									g.write('*(')
+								if i == 0 {
+									if !is_option_unwrap {
+										g.write('*(')
+									}
+									styp := g.base_type(node.obj.typ)
+									g.write('*(${styp}*)')
 								}
-								styp := g.base_type(node.obj.typ)
-								g.write('*(${styp}*)')
 							} else if !g.arg_no_auto_deref {
 								g.write('*')
 							}
@@ -5082,7 +5090,9 @@ fn (mut g Gen) ident(node ast.Ident) {
 								g.write('${dot}_${sym.cname}')
 							} else {
 								if is_option && !node.obj.is_unwrapped {
-									g.write('.data')
+									if i == 0 {
+										g.write('.data')
+									}
 									if !is_option_unwrap {
 										g.write(')')
 									}

--- a/vlib/v/tests/options/option_nested_unwrapping_test.v
+++ b/vlib/v/tests/options/option_nested_unwrapping_test.v
@@ -1,0 +1,20 @@
+type Foo = string | int | f32
+
+fn bar(log ?Foo) {
+	if log != none {
+		dump(log)
+		assert typeof(log).name == 'Foo'
+		if log is string {
+			dump(log)
+			assert true
+			return
+		} else {
+			assert false
+		}
+	}
+	assert false
+}
+
+fn test_main() {
+	bar('foo')
+}

